src/util/piped_process.cpp

```diff
@@ -0,0 +1,361 @@
+    // Close pipes to be used by the child process
+    close(pipe_input[0]);
+    close(pipe_output[1]);
+    // Get stream for sending to the child process
+    command_stream = fdopen(pipe_input[1], "w");
+  }
+#endif
+  process_state = statet::CREATED;
+}
+
+piped_processt::~piped_processt()
+{
+  // Close the parent side of the remaining pipes
+  // and kill child process
+#ifdef _WIN32
+  CloseHandle(child_std_OUT_Rd);
+  CloseHandle(child_std_IN_Wr);
+  TerminateProcess(proc_info.hProcess, 0);
+  CloseHandle(proc_info.hProcess);
+  CloseHandle(proc_info.hThread);
+#else
+  fclose(command_stream);
+  // Note that the above will call close(pipe_input[1]);
+  close(pipe_output[0]);
+  kill(pid, SIGTERM);
+#endif
+}
+
+piped_processt::send_responset piped_processt::send(const std::string &message)
+{
+  if(process_state != statet::CREATED)
+  {
+    return send_responset::ERRORED;
+  }
+#ifdef _WIN32
+  if(!WriteFile(
+    child_std_IN_Wr, message.c_str(), message.size(), NULL, NULL))
+  {
+    // Error handling with GetLastError ?
+    return send_responset::FAILED;
+  }
+#else
+  int send_status = fputs(message.c_str(), command_stream);
+  fflush(command_stream);
+  if(send_status == EOF)
+  {
+    return send_responset::FAILED;
+  }
+#endif
+  return send_responset::SUCCEEDED;
+}
+
+std::string piped_processt::receive()
+{
+  INVARIANT(
+    process_state == statet::CREATED,
+    "Can only receive() from a fully initialised process");
+  std::string response = std::string("");
+  char buff[BUFSIZE];
+  bool success = true;
+#ifdef _WIN32
+  DWORD nbytes;
+#else
+  int nbytes;
+#endif
+  while(success)
+  {
+#ifdef _WIN32
+    success = ReadFile(child_std_OUT_Rd, buff, BUFSIZE, &nbytes, NULL);
+#else
+    nbytes = read(pipe_output[0], buff, BUFSIZE);
+    success = nbytes > 0;   // 0 is error, -1 is nothing to read
+#endif
+    if(nbytes > 0)
+    {
+      response.append(buff, nbytes);
+    }
+  }
+  return response;
+}
+
+std::string piped_processt::wait_receive()
+{
+  // can_receive(PIPED_PROCESS_INFINITE_TIMEOUT) waits an ubounded time until
+  // there is some data
+  can_receive(PIPED_PROCESS_INFINITE_TIMEOUT);
+  return receive();
+}
+
+piped_processt::statet piped_processt::get_status()
+{
+  return process_state;
+}
+
+bool piped_processt::can_receive(optionalt<std::size_t> wait_time)
+{
+  // unwrap the optional argument here
+  int timeout;
+  if(wait_time.has_value())
+  {
+    timeout = wait_time.value();
+  }
+  else
+  {
+    timeout = -1;
+  }
+#ifdef _WIN32
+  int waited_time = 0;
+  // The next four need to be initialised for compiler warnings
+  DWORD buffer = 0;
+  LPDWORD nbytes = 0;
+  LPDWORD rbytes = 0;
+  LPDWORD rmbytes = 0;
+  while(timeout < 0 || waited_time >= timeout)
+  {
+    PeekNamedPipe(child_std_OUT_Rd, &buffer, 1, nbytes, rbytes, rmbytes);
+    if(buffer != 0)
+    {
+      return true;
+    }
+// TODO, make this define and choice better
+#  define WIN_POLL_WAIT 10
+    Sleep(WIN_POLL_WAIT);
```

Comment:
There is no better option than polling? This will heat a lot of CPUs!

Reply:
I haven't found evidence of one that works nicely with our asynchronous reading from named pipes. The options available for waiting appear to work with blocking pipes, but this does not work with our interactive approach (at least at the moment).

If someone who knows Windows better can suggest an alternative I'll look into it.

Reply:
I'd say this is a blocker for using this.

Reply:
`MsgWaitForMultipleObjectsEx` might do the right thing.

Reply:
`WaitForMultipleObjects` (and similar ones) rely upon the events they wait on being synchronous. This (so far in testing and from what I've read) means we have to wait for the child process to finish before we can read from the pipe. I saw some example code that used this approach with asynchronous pipes, but then used a busy loop with `peek` in the same way I've implemented currently.

I'll keep looking, and play around more now I've sorted out most of the other quirks.

Reply:
See comment/discussion at the top of the file for more on why this is done this way.